File: 2021/archived/Day11/Day11.fs
module Day11

open System.IO

let (|EmptySet|_|) a = if Set.isEmpty a then Some () else None

let (<&>) f g x = f x && g x

let (<.>) f x y = f y x

let locate (o: 'a [,]) (x, y) = o.[y, x]

let flashed = snd
let energy = fst

let X = 9
let Y = 9

let inBound (x, y) = not <| (x < 0 || x > X || y < 0 || y > Y)

let neighbors (x, y) =
    [| (0, -1); (0, 1); (1, 0); (-1, 0); (-1, -1); (-1, 1); (1, -1); (1, 1) |]
    |> Array.map (fun (vx, vy) -> x + vx, y + vy)
    |> Array.filter inBound

let flash ocean =
    let waitingToFlash ocean =
        [| for x in 0 .. X do
               for y in 0 .. Y -> x, y |]
        |> Array.filter (
            locate ocean
            >> ((energy >> (>) <.> 9) <&> (not << flashed))
        )
        |> Set.ofArray

    let rec flashing (toBeFlashed: (int * int) Set) ocean =
        match toBeFlashed with
        | EmptySet -> ocean
        | toBeFlashed ->
            let neighbors = toBeFlashed |> Set.map neighbors |> Set.toArray |> Array.concat
            let ocean =
                ocean
                |> Array2D.mapi (fun y x (e, f) ->
                                    if toBeFlashed |> Set.contains (x, y) then e, true
                                    else
                                        let energy = neighbors
                                                    |> Array.filter ((=) (x, y)) |> Array.length
                                        e + energy, f)
            flashing (waitingToFlash ocean) ocean

    let ocean = ocean |> Array2D.map (fun (e, f) -> e + 1, f)

    flashing (waitingToFlash ocean) ocean
    |> Array2D.map (fun (e, _) -> if e > 9 then 0, false else e, false)

let countFlash ocean =
    [| for x in 0 .. X do
        for y in 0 .. Y -> x, y |]
    |> Array.filter (locate ocean >> fst >> (=) 0)
    |> Array.length

let firsrPart ocean =
    [| 1 .. 100 |]
    |> Array.fold (fun (c, o) _ -> (c + countFlash o, flash o)) (0, flash ocean)
    |> fst

let secondPart ocean =
    let rec loop turn ocean =
        let count = countFlash ocean
        if count = (X + 1) * (Y + 1) then turn
        else loop (turn + 1) (flash ocean)

    loop 1 (flash ocean)

let input =
    File.ReadAllLines "input.txt"
    |> Array.map (Seq.toArray >> Array.map (string >> int))
    |> Array.map (fun row -> row |> Array.map (fun cell -> (cell, false)))
    |> array2D

input |> firsrPart |> printfn "First part: %d"
input |> secondPart |> printfn "Second part: %d"

File: 2021/archived/Day11/Day11.scala
import scala.io.Source
import scala.annotation.tailrec

@main def day11 =
    type Point = (Int, Int)
    type Ocean = Vector[Vector[(Int, Boolean)]]

    extension (ocean: Ocean) def |=>(cell: Point) = ocean(cell._2)(cell._1)

    val X = 9
    val Y = 9

    def neighbors(cell: Point) =
        def move(from: Point)(dir: Point) =
            val (x, y) = from
            val (vx, vy) = dir
            (x + vx, y + vy)
        def inBound(cell: Point) =
            val (x, y) = cell
            !(x < 0 || x > X || y < 0 || y > Y)

        Vector(
          (0, 1),
          (0, -1),
          (1, 0),
          (-1, 0),
          (1, 1),
          (1, -1),
          (-1, 1),
          (-1, -1)
        )
            .map(move(cell))
            .filter(inBound)

    def flash(ocean: Ocean) =
        def waitingToFlash(ocean: Ocean) =
            (for x <- 0 to X
            yield for y <- 0 to Y
            yield (x, y))
                .flatMap(identity)
                .filter(cell =>
                    val (energy, flashed) = ocean |=> cell
                    energy > 9 && !flashed
                )
                .toSet

        @tailrec
        def flashing(toBeFlashed: Set[Point], ocean: Ocean): Ocean =
            toBeFlashed match
                case s if s.isEmpty => ocean
                case toBeFlashed =>
                    val nei =
                        toBeFlashed.map(neighbors(_)).toVector.flatMap(identity)

                    val next = ocean.zipWithIndex
                        .map((row, y) =>
                            row.zipWithIndex
                                .map((cell, x) =>
                                    val (e, f) = cell
                                    if toBeFlashed.contains((x, y)) then
                                        (e, true)
                                    else
                                        val plus = nei.count(_ == (x, y))
                                        (e + plus, f)
                                )
                        )
                    flashing(waitingToFlash(next), next)

        val flashed = ocean.map(_.map(cell => (cell._1 + 1, cell._2)))

        flashing(waitingToFlash(flashed), flashed)
            .map(_.map({ case (e, _) =>
                if e > 9 then (0, false) else (e, false)
            }))

    def countFlash(ocean: Ocean) =
        (for x <- 0 to X
        yield for y <- 0 to Y
        yield (x, y))
            .flatMap(identity)
            .count(cell => (ocean |=> cell)._1 == 0)

    def firstPart(ocean: Ocean) =
        (1 to 100)
            .foldLeft((0, flash(ocean)))((state, _) =>
                val (count, ocean) = state
                (count + countFlash(ocean), flash(ocean))
            )
            ._1

    def secondPart(ocean: Ocean) =
        @tailrec
        def loop(turn: Int, ocean: Ocean): Int =
            val count = countFlash(ocean)
            if count == (X + 1) * (Y + 1) then turn
            else loop((turn + 1), flash(ocean))

        loop(1, flash(ocean))

    val input = Source
        .fromFile("input.txt")
        .getLines
        .map(_.toCharArray.toVector.map(c => (c.toString.toInt, false)))
        .toVector

    println(s"First part: ${firstPart(input)}")
    println(s"Second part: ${secondPart(input)}")
